### ISSUE TITLE
until with Runnable version that allows throwing exception

### ISSUE THREAD

**Issue by tkowalcz** [jayway/awaitility#32]
Consider providing until version that accepts lambda expression that is able to throw:

```java
Awaitility.await().atMost(...).until(() -> {
...
callThatThrowsException();
...
})
````

Right now this looks like:

```java
Awaitility.await().atMost(...).until(() -> {
...
try {
    callThatThrowsException();
} catch(Exception e) {
    throw new RuntimeException(e);
}
...
})
````

I can provide an implementation if you are ok with until() accepting something like `ThrowableRunnable`.

**Comment by johanhaleby** [comment]
Thanks for reporting. Is it possible to do this without breaking backward compatibility?

**Comment by szpak** [comment]
Good point Johan. Overloaded `unitl` methods accepting both `Runnable` and anything else having a single `void xxx()` method would clash on lambda expressions. For case mentioned by Tomek `Callable<Void>` could be almost used (as `call` method declares `Exception`), but `callThatThrowsException` would have to return `Void` instead of `void` (and in addition a method `until` with `Callable<Boolean>` is already used).

As a result I don't know if an another method name could be avoided.

Looking at the possible implementation `AssertionCondition` would have to refactored to support both `Runnable` and `ThrowableRunnable` (they do not have a common interface to create typed class) to not duplicate code. One of the possible solutions could be switching `AssertionCondition` to use `ThrowableRunnable` (instead `Runnable`) and inline wrapping `Runnable` (from `until(Runnable)`) with just `ThrowableRunnable`.

**Comment by tkowalcz** [comment]
I did a prototype implementation (which is fairly easy), but the problem is exactly backward compatibility.

Introducing until(ThrowingRunnable) alongside until(Runnable) breaks compilation (e.g. visible also in awaitility tests compilation problems). Also it requires casting of lambdas to inform the compiler which overload to use.

I'm open for ideas :) I'll let you know when I find a solution.

**Comment by tkowalcz** [comment]
Can someone look at my proposal to verify that it is the right way to go? We would then might work out the method naming thing :)

**Comment by szpak** [comment]
I've seen it, but I was waiting for a PR (thinking that you were planning something else to work on).

**Comment by tkowalcz** [comment]
I guess the last thing is to find a name :)

**Comment by johanhaleby** [comment]
Thanks for helping out. I'll look into this asap.

**Comment by johanhaleby** [comment]
Please have a look at the [pull request](https://github.com/jayway/awaitility/pull/35) and give me feedback.

**Comment by bcalmac** [comment]
This is already closed, but I would still try some feedback. 

From an API usage point of view, the matches() approach is nicer than having a try/catch in the lambda, but when you work on a code base with ubiquitous checked exceptions, it starts feeling like unnecessary boilerplate and affects readability:

```java
await().untilAsserted(matches(this::myAssertionMethod));
```
as opposed to the direct alternative:

```java
await().untilXxx(this::myAssertionMethod);
```

@tkowalcz has in his fork a untilPass() method implementing the second approach. Thoughts?

**Comment by tkowalcz** [comment]
IIRC the main driver for forcing `matches` here is that my approach breaks existing code using the API. You can't have untilAsserted that accept Runnable as well as ThrowingRunnable.

**Comment by bcalmac** [comment]
It can be a different name.

**Comment by johanhaleby** [comment]
@bcalmac Thanks for you input. The problem I see with `untilXXX` in this case is that I would have to create (at least) two new methods for `matches`. One `untilCheckedAsserted` (that takes `ThrowingRunnable` and `untilChecked(ThrowingRunnable)`. Maybe there are other overloaded methods as well that I'm not thinking of off the top of my head. I (currently) see `matches` more as a convenience function to deal with checked exceptions.  But I agree that `await().untilAsserted(matches(this::myAssertionMethod));` is not super-nice.

**Comment by bcalmac** [comment]
The existing API has only one method taking Runnable, so there would be just one new method taking `ThrowingRunnable`, let's call it `untilCheckedAsserted()`. 

The `until()` method has a very different semantic (takes a predicate, not a block of assertions) and since it uses `Callable`, it's already fine with checked exceptions.

**Comment by johanhaleby** [comment]
True.. I was first thinking that `matches` also worked for cases like the one below where `something` would throw a checked exception:

```java
await().until(() -> something("2"), equalTo("4"));
```

But it's not required as you say.

Hmm interesting :) I would agree that `untilCheckedAsserted` might read better than `untilAsserted(matches(...))`. WDYT @tkowalcz?

**Comment by bcalmac** [comment]
The naming is probably the hardest part :-) 

`untilCheckedAsserted()` is consistent but it sounds very confusing. I would go with something like`untilAssertedEx()` or `untilAssertedX()` to express in a very minimal way that this is a variant that can accept lambdas throwing (checked) exceptions.

**Comment by johanhaleby** [comment]
Yeah good point. But what should `X` be? :)

**Comment by bcalmac** [comment]
Oh, I meant a literal X. That would be the name.

**Comment by johanhaleby** [comment]
Oh :) I personally don't find X to be a very descriptive name, I'd prefer something else.

**Comment by bcalmac** [comment]
Ex (from exception)? If not, maybe `untilThrowingAsserted()` consistent with `ThrowingRunnable`.

And when everything else fails, there's always `whenAsserted42()`, of course :-)

**Comment by johanhaleby** [comment]
Hmmm yeah naming is always hard.. Maybe `untilThrowingAsserted` might be confusing as well since you can indeed throw exceptions (not just checked exceptions) from the normal `untilAsserted` function.

**Comment by johanhaleby** [comment]
Might be a crazy idea, but would about getting rid of `untilAsserted(Runnable)` and only have `untilAsserted(ThrowableRunnable)`? Would that work? If using lambda expressions I suppose it shouldn't make a difference?

**Comment by bcalmac** [comment]
That would be ideal. It would only be backward incompatible for clients using actual Runnable instances as an argument (as opposed to lambdas).

Considering that 3.0 is almost out, now it's a good time to introduce this slightly backward incompatible change.

**Comment by johanhaleby** [comment]
I've actually implemented this in master now and published a new snapshot. Please try version `3.0.0-SNAPSHOT` after having added the following maven repo:

```xml
<repositories>
        <repository>
            <id>sonatype</id>
            <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
            <snapshots />
        </repository>
</repositories>
```

My plan was to release 3.0.0 tomorrow but I'll probably release another release candidate instead to give other people time to test it.

**Comment by bcalmac** [comment]
It works, but more importantly it looks beautiful! Thank you for listening.

**Comment by johanhaleby** [comment]
Cool! Thanks for trying it out and for your feedback. I hadn't thought about this. And it was excellent timing due to the upcoming 3.0 release. I'll release a new rc today.

**Comment by johanhaleby** [comment]
I've now released 3.0.0-rc3
